trl/trainer/grpo_trainer.py

```diff
@@ -1402,17 +1418,54 @@ def _generate_and_score_completions(
         # Apply weights to each reward function's output and sum
         rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
 
+        num_in_group = self.num_generations
+        num_inputs_in_device = len(prompts)
+
+        if self.num_remains_in_group is not None and mode == "train":
+            num_in_group = self.num_remains_in_group
+
+            all_completions = gather_object(completions)
+
+            group_filter_scores = self.group_filter_func(
+                group_completions=[
+                    all_completions[i : i + 1 * self.num_generations]
+                    for i in range(len(all_completions) // self.num_generations)
+                ],
+                group_rewards=rewards.view(-1, self.num_generations).tolist(),
+            )
+            group_filter_scores = torch.tensor(
+                group_filter_scores, device=device
+            )
+
+            _, group_local_indices = torch.topk(
+                group_filter_scores, self.num_remains_in_group, dim=-1
+            )
+            group_row_offsets = torch.arange(
+                0, len(all_completions), self.num_generations, device=device
+            ).unsqueeze(1)
+            group_global_indices = group_row_offsets + group_local_indices
+            group_global_indices = group_global_indices.flatten()
+
+            rewards = rewards[group_global_indices].contiguous()
+            rewards_per_func = rewards_per_func[
+                group_global_indices, :
+            ].contiguous()
+
+            num_inputs_in_device = int(
+                len(prompts) / self.num_generations * self.num_remains_in_group
+            )
+
         # Compute grouped-wise rewards
-        mean_grouped_rewards = rewards.view(-1, self.num_generations).mean(dim=1)
+        mean_grouped_rewards = rewards.view(-1, num_in_group).mean(dim=1)
 
         # Normalize the rewards to compute the advantages
-        mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(self.num_generations, dim=0)
+        mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(num_in_group, dim=0)
         advantages = rewards - mean_grouped_rewards
 
         if self.scale_rewards in ["batch", "none"]:
```

Comment:
It seems this branch is doing scale_rewards in "group" senario, should I change this in this PR?